src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,244 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Taken from https://github.com/dotnet/aspnetcore/blob/master/src/Mvc/Mvc.Core/src/Formatters/TranscodingReadStream.cs
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    internal sealed class TranscodingReadStream : Stream
+    {
+        private static readonly int OverflowBufferSize = Encoding.UTF8.GetMaxByteCount(1); // The most number of bytes used to represent a single UTF char
+
+        internal const int MaxByteBufferSize = 4096;
+        internal const int MaxCharBufferSize = 3 * MaxByteBufferSize;
+
+        private readonly Stream _stream;
+        private readonly Decoder _decoder;
+        private readonly Encoder _encoder;
+
+        private ArraySegment<byte> _byteBuffer;
+        private ArraySegment<char> _charBuffer;
+        private ArraySegment<byte> _overflowBuffer;
+        private bool _disposed;
+
+        public TranscodingReadStream(Stream input, Encoding sourceEncoding)
+        {
+            _stream = input;
+
+            // The "count" in the buffer is the size of any content from a previous read.
+            // Initialize them to 0 since nothing has been read so far.
+            _byteBuffer = new ArraySegment<byte>(ArrayPool<byte>.Shared.Rent(MaxByteBufferSize), 0, count: 0);
+
+            // Attempt to allocate a char buffer than can tolerate the worst-case scenario for this
+            // encoding. This would allow the byte -> char conversion to complete in a single call.
+            // However limit the buffer size to prevent an encoding that has a very poor worst-case scenario.
+            // The conversion process is tolerant of char buffer that is not large enough to convert all the bytes at once.
+            int maxCharBufferSize = Math.Min(MaxCharBufferSize, sourceEncoding.GetMaxCharCount(MaxByteBufferSize));
```

Comment:
_byteBuffer might be significantly larger than MaxByteBufferSize, as the ArrayPool can hand out larger arrays than asked for.  But this computation is based off of MaxByteBufferSize rather than the actual array length.  Is that ok?

Reply:
We should be fine since we copy leftover bytes on _byteBuffer to the beginning of the array before reading again from the underlying stream.
https://github.com/dotnet/runtime/pull/33459/files#diff-8b4bda2579bb0d5dfbf6b955a958d384R182

Is there a way to test this scenario? I was trying to exhaust the ArrayPool by just renting a bunch of times until a larger array was returned but I never got a larger buffer by just calling `ArrayPool<byte>.Shared.Rent(MaxBufferSize)`.

This is what I am doing:
```cs
[Fact]
public async Task ByteBufferIsLargerThanCharBuffer()
{
    List<byte[]> rentedArrays = new List<byte[]>();

    while (true)
    {
        byte[] rented = ArrayPool<byte>.Shared.Rent(TranscodingReadStream.MaxByteBufferSize);

        rentedArrays.Add(rented);
        if (rented.Length > TranscodingReadStream.MaxByteBufferSize)
        {
            // Rented array is larger than expected... proceeding with the test.
            break;
        }
    }

    /* Run test */

    /* Return rented arrays */
}
```

Reply:
The implementation of ArrayPool.Shared currently says that it'll always give back power-of-two arrays.  Since you asked for 4096, you always get the right size.  Ask for 4097 :smile:

Reply:
Having a smaller char buffer isn't incorrect - the decoder will perform multiple partial conversions until the byte array is exhausted. However changing this to `_byteBuffer.Length` might result in fewer partial decodings since the char buffer would be large enough to accommodate decoding all the bytes in a single call to `Convert`.

Reply:
@bartonjs What I meant is, how can I get 8192 while still asking for 4096?

Reply:
> What I meant was, how can I get 8192 while still asking for 4096?

With the current implementation of the shared array pool, you can't. (It's really more like 10 different specific-sized array pools.)  But if we eliminated the 4096 bucket then you'd always get 8192.  Or if we made local-pool exhaustion rent from the next bigger one, then you might get back a megabyte buffer that was just sitting idle.  Neither of these things would violate the documented behavior, which is that the array is at least as big as the size requested.